File: implementations/dart/lib/src/move.dart
import 'package:chess_engine/chess_engine.dart';

class Move {
  final int fromRow;
  final int fromCol;
  final int toRow;
  final int toCol;
  final PieceType? promotion;

  Move(this.fromRow, this.fromCol, this.toRow, this.toCol, {this.promotion});

  @override
  String toString() {
    return '${_colToChar(fromCol)}${8 - fromRow}${_colToChar(toCol)}${8 - toRow}${promotion != null ? _typeToChar(promotion!) : ''}';
  }

  static String _colToChar(int col) {
    return String.fromCharCode('a'.codeUnitAt(0) + col);
  }

  static String _typeToChar(PieceType type) {
    switch (type) {
      case PieceType.queen:
        return 'q';
      case PieceType.rook:
        return 'r';
      case PieceType.bishop:
        return 'b';
      case PieceType.knight:
        return 'n';
      default:
        return '';
    }
  }
}

File: implementations/java/src/main/java/Move.java
public class Move {
    private int fromRow;
    private int fromCol;
    private int toRow;
    private int toCol;
    private PieceType promotionPiece;
    private Piece capturedPiece;
    private boolean isCastling;
    private boolean isEnPassant;
    private int previousEnPassantCol;
    private boolean[] previousCastlingRights;

    public Move(int fromRow, int fromCol, int toRow, int toCol) {
        this.fromRow = fromRow;
        this.fromCol = fromCol;
        this.toRow = toRow;
        this.toCol = toCol;
        this.promotionPiece = null;
        this.capturedPiece = null;
        this.isCastling = false;
        this.isEnPassant = false;
        this.previousEnPassantCol = -1;
        this.previousCastlingRights = new boolean[4];
    }

    public int getFromRow() { return fromRow; }
    public int getFromCol() { return fromCol; }
    public int getToRow() { return toRow; }
    public int getToCol() { return toCol; }
    public PieceType getPromotionPiece() { return promotionPiece; }
    public void setPromotionPiece(PieceType piece) { this.promotionPiece = piece; }
    public Piece getCapturedPiece() { return capturedPiece; }
    public void setCapturedPiece(Piece piece) { this.capturedPiece = piece; }
    public boolean isCastling() { return isCastling; }
    public void setCastling(boolean castling) { this.isCastling = castling; }
    public boolean isEnPassant() { return isEnPassant; }
    public void setEnPassant(boolean enPassant) { this.isEnPassant = enPassant; }
    public int getPreviousEnPassantCol() { return previousEnPassantCol; }
    public void setPreviousEnPassantCol(int col) { this.previousEnPassantCol = col; }
    public boolean[] getPreviousCastlingRights() { return previousCastlingRights; }
    public void setPreviousCastlingRights(boolean[] rights) {
        System.arraycopy(rights, 0, this.previousCastlingRights, 0, 4);
    }

    @Override
    public String toString() {
        String from = "" + (char)('a' + fromCol) + (8 - fromRow);
        String to = "" + (char)('a' + toCol) + (8 - toRow);
        if (promotionPiece != null) {
            return from + to + promotionPiece.getSymbol();
        }
        return from + to;
    }
}
